### ISSUE TITLE
Is it possible to construct a rule to verify that an async method is awaited if it has an [MustBeAwaited] attribute?

### ISSUE THREAD

**Issue by JorgenLindell** [meziantou/Meziantou.Analyzer#207]
Sometimes the logic of an async method assumes that it must be awaited when called, Maybe due to that a parameter later in the flow would be destroyed or in another state than at the call. 

By adding an attribute on the method, like [MustBeAwaited] I would like to enforce that the method always is called with await, not just as a Task returning method. 

An example where this would be useful is for example https://asherman.io/projects/csharp-async-reentrant-locks.html
where his REAsyncLock.DoWithLock method always need to be called with await to avoid deadlocks

**Comment by meziantou** [comment]
Can you provide a small demo that shows the cases the analyzer should detect and the expected fixes?
In some cases, Visual Studio already provides a message for non-awaited tasks. I just want to be sure this new proposal won't be a duplicate of an existing rule.

Also, what would be the namespace for this attribute? Does JetBrains annotations have something similar?

**Comment by JorgenLindell** [comment]
Here is a demo:

```
using System;
using System.Diagnostics;
using System.Threading.Tasks;

namespace TestAsyncAwait
{
  public class Program
  {
    [AttributeUsage(AttributeTargets.Method)]
    private class MustBeAwaitedAttribute : Attribute { } //dummy implementation of attribute
   
    private class ParamClass
    {
      public int Value;
    }

    private static async Task A1()
    {
      await Task.Delay(TimeSpan.FromSeconds(1));
    }

    [MustBeAwaited]
    private static async Task A2()
    {
      await Task.Delay(TimeSpan.FromSeconds(1));
    }

    [MustBeAwaited]
    private static async Task<int> F2(ParamClass p)
    {
      await Task.Delay(TimeSpan.FromSeconds(1));
      return p.Value;
    }

    static async Task Main(string[] args)
    {
      A1(); // gives IDE0058 CS4014, which are warnings, only CS4014 is related to async, but is a general warning, just recommending await

      Task runLater1 = A1(); //No issues in VS with ReSharper and SonarLint installed
      Task runLater2 = A1();

      await Task.WhenAll(runLater1, runLater2);

      A2(); // should be an error

      runLater1 = A2();  // should be an error
      runLater2 = A2(); // should be an error
      await A2(); // Correct usage
      await Task.WhenAll(runLater1, runLater2);

      ParamClass param = new ParamClass
      {
        Value = 3
      };
      var correctInt = await F2(param); // correct call

      var intTask = F2(param); // should be an error

      param.Value = 4;// code here that changes state of param to something not expected by F2
      int possiblyBadInt = await intTask;

      Debug.WriteLine($"correctInt={correctInt}");
      Debug.WriteLine($"possiblyBadInt={possiblyBadInt}");

    }
  }
}

```
CS4014 is a bit like this, but a) it is just a warning normally, and b) I would like to explicitly regard it as an error when occuring together with that attribute. Also, it seems like CS4014 isn't triggered by  Task runLater2 = A1(); which I thought it should.

Regarding what namespace I'm not sure I understand the question. All other things async/await is in System.Threading.Tasks, is that an answer? :) Other analyzer packages seem to have the attributes as part of the analyzer package or as a separate nuget package.

**Comment by meziantou** [comment]
I look at multiple projects and cannot see any actual use case for this rule, but maybe you use patterns I don't use. That's why I'm asking for more details.

> CS4014 is a bit like this, but a) it is just a warning normally

People should use WarningsAsErrors at least when using the "Release" configuration ([example](https://github.com/meziantou/Meziantou.Framework/blob/0514badad0633caea60841a69d6aed6a0ddb9a20/Directory.Build.props#L8)). Warnings are important and should be fixed. So, this case doesn't need to be handled by the rule.

> Also, it seems like CS4014 isn't triggered by Task runLater2 = A1(); which I thought it should

The compiler understand you know that the call returns a task, so you have handled it _correctly_. If you do not await it later, it's ok from a compiler point of view. More info in the documentation: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs4014

> `param.Value = 4;// code here that changes state of param to something not expected by F2`

Indeed. Is it a theorical example, or have you seen this case in actual code?

I think that if a method could be called in parallel, it should clone its parameters, or use immutable types, or use another pattern that prevents this kind of issues.

> where his REAsyncLock.DoWithLock method always need to be called with await to avoid deadlocks

There are other ways to get the same logic and avoid the non-awaited task issue. For instance, it could return an IDisposable when the lock is taken. In this case, you must await the call to get the disposable instance and use it in a `using` block. If you forget the await, the code won't compile. (this is just one case, maybe you have other examples)

### Implementation details (if this rule is approved)

The analyzer should register on `IMethodInvokeOperation` and check if the operation has a parent of type `IAwaitOperation`. However, there are a few exceptions:
- `F().Wait()` / `F().Result` / `F().GetAwaiter().GetResult()` are valid as the task is _implicitly_ awaited
- `await Task.WhenAll(F(), F())` is not valid if I well understood the issue while there are a parent await statement
- `await F().ConfigureAwait(false)` is valid while the immediate parent is not an await operation
- `await foreach(var a in F())` is valid while there is no await expession (need to check `ILoopOperation.IsAsynchronous`)
- `F().ContinueWith(...)` is not valid

I'm not sure what's the best way to handle all these cases correctly. Maybe the list of allowed methods should be hardcoded.

**Comment by JorgenLindell** [comment]
Actually it was the note on REAsyncLock.DoWithLock that had me thinking about this. I will consider refactoring that solution to use IDisposable as you mention and see if I can get that to work in a way that effectively stops that issue to occur.

The complexities of this rule is a bit out of my leauge, I am not that much into compilers and analysers to really understand that.

As I understand it, the need for this kind of rule in the case of REAsyncLock arises from how he use AsyncLocal<> and that this makes the internal state of the object dependent on the async call-order, Which can be broken if calls to it are not awaited.

Without any other real life example, I still kind of assume that this could be a more general, although maybe rare, case, that might be relevant to catch explicitly.

Anyway, I will consider that lead on returning an IDisposable to see if that can solve the issue in that specific case in a reasonable way.

Thanks for spending time considering this!

**Comment by meziantou** [comment]
I'll close the issue as the suggested rule seems to apply to a very specific use case. Feel free to reopen it if you have a use case with a method from the .NET Framework or a popular library.
